### PR TITLE
improved mongodb query formatting

### PR DESCRIPTION
**Pull request by WebDevMichael**
- Prevent from using redundant $and
- Improved 'between', 'notInBetween'
- Remove redundant '$eq' operator

### CODE REVIEW

**Comment by WebDevMichael** [comment]
I'm struggling with the RuleType typings, that's what I' always getting stuck at 😄

File: .gitignore [hunk 1]
```diff
@@ -8,3 +8,6 @@ dist
 /types
 .parcel-cache
 .env
+/.codesandbox/
+/.github/
+/.vscode/
```
Comment (jakeboone02) [inline .gitignore:13]: Why was this file modified?

File: packages/react-querybuilder/package.json [hunk 2]
```diff
@@ -59,6 +59,7 @@
     "vite-tsconfig-paths": "^3.3.17"
   },
   "dependencies": {
+    "bson": "^4.6.4",
```
Comment (jakeboone02) [inline packages/react-querybuilder/package.json:62]: Why was this package added? Where it's being used, it looks like you could just use `JSON.parse` instead of `EJSON.parse`. Even if `bson` was necessary, it should be in  `devDependencies`, not `dependencies`.

File: packages/react-querybuilder/src/utils/formatQuery/defaultValueProcessorMongoDBByRule.ts [hunk 3]
```diff
@@ -17,80 +17,73 @@ export const defaultValueProcessorMongoDBByRule: ValueProcessorByRule = (
...
     operator === '>' ||
     operator === '>='
   ) {
     const mongoOperator = mongoOperators[operator];
-    return valueIsField
-      ? `{"$expr":{"${mongoOperator}":["$${field}","$${value}"]}}`
-      : `{"${field}":{"${mongoOperator}":${
-          useBareValue ? trimIfString(value) : `"${escapeDoubleQuotes(value)}"`
-        }}}`;
+    return `"${field}":{"${mongoOperator}":${
```
Comment (jakeboone02) [inline packages/react-querybuilder/src/utils/formatQuery/defaultValueProcessorMongoDBByRule.ts:25]: Why does this ignore the case where `valueSource === "field"`?

File: packages/react-querybuilder/src/utils/formatQuery/defaultValueProcessorMongoDBByRule.ts [hunk 4]
```diff
@@ -17,80 +17,73 @@ export const defaultValueProcessorMongoDBByRule: ValueProcessorByRule = (
...
+      .map(val =>
+        shouldRenderAsNumber(val, parseNumbers)
+          ? `${trimIfString(val)}`
+          : `"${escapeDoubleQuotes(val)}"`
+      )
+      .join(',')}]}`;
   } else if (operator === 'between' || operator === 'notBetween') {
     const valArray = toArray(value);
-    if (valArray.length >= 2 && isValidValue(valArray[0]) && isValidValue(valArray[1])) {
+    if (isValidValue(valArray[0]) && isValidValue(valArray[1])) {
```
Comment (jakeboone02) [inline packages/react-querybuilder/src/utils/formatQuery/defaultValueProcessorMongoDBByRule.ts:69]: Why was the check for minimum array length of 2 removed?

File: packages/react-querybuilder/src/utils/formatQuery/defaultValueProcessorMongoDBByRule.ts [hunk 5]
```diff
@@ -17,80 +17,73 @@ export const defaultValueProcessorMongoDBByRule: ValueProcessorByRule = (
...
       if (operator === 'between') {
-        return valueIsField
-          ? `{"$and":[{"$expr":{"$gte":["$${field}","$${firstValue}"]}},{"$expr":{"$lte":["$${field}","$${secondValue}"]}}]}`
-          : `{"$and":[{"${field}":{"$gte":${firstValue}}},{"${field}":{"$lte":${secondValue}}}]}`;
+        return `"${field}":{"$gte":${firstValue},"$lte":${secondValue}}`;
       } else {
-        return valueIsField
-          ? `{"$or":[{"$expr":{"$lt":["$${field}","$${firstValue}"]}},{"$expr":{"$gt":["$${field}","$${secondValue}"]}}]}`
-          : `{"$or":[{"${field}":{"$lt":${firstValue}}},{"${field}":{"$gt":${secondValue}}}]}`;
+        return `"${field}":{"$lte":${firstValue},"$gte":${secondValue}}`;
```
Comment (jakeboone02) [inline packages/react-querybuilder/src/utils/formatQuery/defaultValueProcessorMongoDBByRule.ts:86]: More examples where the `valueSource === "field"` case is ignored.

File: packages/react-querybuilder/src/utils/formatQuery/formatQuery.test.ts [hunk 6]
```diff
@@ -630,10 +631,9 @@ const params_named = {
...
+  '{"$and":[{},{"firstName":null},{"lastName":{"$ne":null}},{"firstName":{"$in":["middleName","lastName"]}},{"lastName":{"$nin":["middleName","lastName"]}},{"firstName":{"$in":[]}},{"firstName":{"$gte":"middleName","$lte":"lastName"}},{"firstName":{"$gte":"middleName","$lte":"lastName"}},{"lastName":{"$lte":"middleName","$gte":"lastName"}},{"firstName":{"$gte":"","$lte":""}},{"firstName":{"$gte":"","$lte":""}},{"age":"iq"},{"isMusician":"isCreative"},{"$where":"this.email.includes(this.atSign)"},{"$where":"this.email.startsWith(this.name)"},{"$where":"this.email.endsWith(this.dotCom)"},{"$where":"!this.hello.includes(this.dotCom)"},{"$where":"!this.job.startsWith(this.noJob)"},{"$where":"!this.job.endsWith(this.noJob)"},{"$or":[{"job":"executiveJobName"}]}]}';
```
Comment (jakeboone02) [inline packages/react-querybuilder/src/utils/formatQuery/formatQuery.test.ts:636]: What's with the addition of the empty rule object (first element of the `$and` array)?

**Review by jakeboone02 — COMMENTED** [review]
Why was a `package-lock.json` added? This library uses Yarn for dependency management.

**Comment by jakeboone02** [comment]
Do you have specific questions about `RuleType`? I feel like that one is fairly straightforward. `RuleGroupTypeAny`, however, can get weird since it can be either `RuleGroupType` or `RuleGroupTypeIC`.

File: .gitignore (hunk 1 above)
Comment (WebDevMichael) [reply .gitignore:13]: Because I believe we do not need these folders in Git

File: packages/react-querybuilder/package.json (hunk 2 above)
Comment (WebDevMichael) [reply packages/react-querybuilder/package.json:62]: For this PR it is not needed, i will remove it 👍🏻

File: packages/react-querybuilder/src/utils/formatQuery/defaultValueProcessorMongoDBByRule.ts (hunk 3 above)
Comment (WebDevMichael) [reply packages/react-querybuilder/src/utils/formatQuery/defaultValueProcessorMongoDBByRule.ts:25]: Sorry, this valueSource is not clear for me.
`valueSource === "field"` What is the use case for this ? When is the querybuilder generating this?

File: packages/react-querybuilder/src/utils/formatQuery/defaultValueProcessorMongoDBByRule.ts (hunk 4 above)
Comment (WebDevMichael) [reply packages/react-querybuilder/src/utils/formatQuery/defaultValueProcessorMongoDBByRule.ts:69]: We are preparing a between, if there are no 2 valid values, then it's not a valid query. I personally need this code change for my use case. This code change will reflect the current state of the querybuilder as a visual mongo query, in my use case.

File: packages/react-querybuilder/src/utils/formatQuery/formatQuery.test.ts (hunk 6 above)
Comment (WebDevMichael) [reply packages/react-querybuilder/src/utils/formatQuery/formatQuery.test.ts:636]: `{
      field: 'invalid',
      operator: 'invalid',
      value: '',
      valueSource: 'field',
    }`
It's for a rule with an invalid operator, how should we handle this, i was thinking of showing it in the query, indicating that it's an invalid rule. The user can then still remove the invalid rule or make the rule valid

**Comment by WebDevMichael** [comment]
> Do you have specific questions about `RuleType`? I feel like that one is fairly straightforward. `RuleGroupTypeAny`, however, can get weird since it can be either `RuleGroupType` or `RuleGroupTypeIC`.

`formatQuery.ts(348,20): error TS2339: Property 'field' does not exist on type 'RuleType<string, string, any> | RuleGroupType<RuleType<string, string, any>, string>'.`

This error is very confusing, why is this 'field' not recognized, it should result in type 'string'

**Comment by WebDevMichael** [comment]
> Why was a `package-lock.json` added? This library uses Yarn for dependency management.

I just removed package-lock.json file ✔️

File: .gitignore (hunk 1 above)
Comment (jakeboone02) [reply .gitignore:13]: We _absolutely_ need these folders in the repo. They have CI configurations and suggested VSCode settings/extensions.

File: packages/react-querybuilder/src/utils/formatQuery/defaultValueProcessorMongoDBByRule.ts (hunk 3 above)
Comment (jakeboone02) [reply packages/react-querybuilder/src/utils/formatQuery/defaultValueProcessorMongoDBByRule.ts:25]: It's for comparing a field to another field, instead of a field to a value. [See documentation here](https://react-querybuilder.js.org/docs/api/valueeditor#value-sources).

File: packages/react-querybuilder/src/utils/formatQuery/defaultValueProcessorMongoDBByRule.ts (hunk 4 above)
Comment (jakeboone02) [reply packages/react-querybuilder/src/utils/formatQuery/defaultValueProcessorMongoDBByRule.ts:69]: It needs to stay. It may seem redundant, but it allows it to immediately bail on processing the rule if the array length is less than 2.

File: packages/react-querybuilder/src/utils/formatQuery/formatQuery.test.ts (hunk 6 above)
Comment (jakeboone02) [reply packages/react-querybuilder/src/utils/formatQuery/formatQuery.test.ts:636]: You should use [validation](https://react-querybuilder.js.org/docs/api/validation) to style invalid rules within the `<QueryBuilder />` component instead of producing invalid output with `formatQuery`.

**Comment by jakeboone02** [comment]
> `formatQuery.ts(348,20): error TS2339: Property 'field' does not exist on type 'RuleType<string, string, any> | RuleGroupType<RuleType<string, string, any>, string>'.`
> 
> This error is very confusing, why is this 'field' not recognized, it should result in type 'string'

It's because `RuleGroupType` is recursive through its `rules` property. `rules` is not simply an array of `RuleType` objects--each element can be a `RuleType` or a nested `RuleGroupType`, so `field` isn't necessarily available. You can use a type guard to let TypeScript know which kind of element you're dealing with:

```ts
if ('rules' in rg) {
  // rg is a RuleGroupType
} else {
  // rg is a RuleType
}
```

**Comment by jakeboone02** [comment]
Closing in favor of #386.
